Ginger/GingerCoreNET/Drivers/CoreDrivers/Web/Playwright/PlaywrightDriver.cs

```diff
@@ -0,0 +1,1037 @@
+                }
+
+                if (_windows.Count > 0)
+                {
+                    _currentWindow = _windows.Last!.Value;
+                }
+                else
+                {
+                    _currentWindow = Task.Run(() =>
+                    {
+                        //this code needs to be executed in a separate Task otherwise, it will cause a deadlock and freeze the calling thread
+                        //check this for an example https://stackoverflow.com/a/43912280/12190808
+                        return NewWindowAsync().Result;
+                    }).Result;
+                }
+            }
+
+            public async Task<IBrowserWindow> NewWindowAsync()
+            {
+                ThrowIfClosed();
+
+                IPlaywrightBrowserContext context = await _playwrightBrowser.NewContextAsync(new BrowserNewContextOptions()
+                {
+                    ViewportSize = ViewportSize.NoViewport,
+                });
+                PlaywrightBrowserWindow window = new(context, OnWindowClose);
+                _windows.AddLast(window);
+                return window;
+            }
+
+            private Task OnWindowClose(IBrowserWindow closedWindow)
+            {
+                _windows.Remove(closedWindow);
+
+                if (_windows.Count <= 0)
+                {
+                    return ClosePlaywrightBrowser();
+                }
+
+                if (closedWindow == _currentWindow)
+                {
+                    _currentWindow = _windows.Last!.Value;
+                }
+
+                return Task.CompletedTask;
+            }
+
+            public Task CloseAsync()
+            {
+                return CloseAllWindows();
+            }
+
+            private Task CloseAllWindows()
+            {
+                Task[] windowCloseTasks = new Task[_windows.Count];
+                List<IBrowserWindow> windowsToClose = new(_windows);
+                for (int index = 0; index < windowsToClose.Count; index++)
+                {
+                    windowCloseTasks[index] = windowsToClose[index].CloseAsync();
+                }
+                return Task.WhenAll(windowCloseTasks);
+            }
+
+            private async Task ClosePlaywrightBrowser()
+            {
+                if (_isClosed)
+                {
+                    return;
+                }
+
+                _isClosed = true;
+
+                await _playwrightBrowser.CloseAsync();
+                await _playwrightBrowser.DisposeAsync();
+                await _onBrowserClose.Invoke(closedBrowser: this);
+            }
+
+            private void ThrowIfClosed()
+            {
+                if (_isClosed)
+                {
+                    throw new InvalidOperationException("Cannot perform operation, browser is already closed.");
+                }
+            }
+        }
+
+        private sealed class PlaywrightBrowserWindow : IBrowserWindow
+        {
```

Comment:
Better to have separate class for PlaywrightBrowserWindow class

Reply:
handled